test/screenshot/screenshot.js

```diff
@@ -1,60 +1,209 @@
+
+      // Save the snapshot and the diff
+      await Promise.all([
+        this.saveImage_(snapshotPath, snapshot, metadata),
+        this.saveImage_(diffPath, diff, metadata),
+      ]);
+
+      return assert.isBelow(Number(data.misMatchPercentage), 0.01);
+    });
+  }
+
+  /**
+   * Generates a unique hash from an image's contents
+   * @param {!Buffer} imageBuffer The image buffer to hash
+   * @return {string}
+   * @private
+   */
+  generateImageHash_(imageBuffer) {
+    return createHash('sha256').update(imageBuffer).digest('hex');
+  }
+
+  /**
+   * Returns the golden hash
+   * @return {string|undefined}
+   * @private
+   */
+  async getGoldenHash_() {
+    const goldenFile = await readFilePromise(goldenFilePath);
+    const goldenJSON = JSON.parse(goldenFile);
+    return goldenJSON[this.urlPath_];
+  }
+
+  /**
+   * Returns the correct image path
+   * @param {string} imageHash The image hash
+   * @param {string} imageType The image type
+   * @return {string|undefined}
+   * @private
+   */
+  getImagePath_(imageHash, imageType) {
+    if (imageType === 'golden') {
+      return `${this.urlPath_}/${imageHash}.golden.png`;
+    }
+
+    if (['snapshot', 'diff'].includes(imageType)) {
+      return `${this.urlPath_}/${commitHash}/${imageHash}.${imageType}.png`;
+    }
+  }
+
+  /**
+   * Downloads an image from Google Cloud Storage
+   * @param {string} gcsFilePath The file path on Google Cloud Storage
+   * @return {Buffer}
+   * @private
+   */
+  async readImage_(gcsFilePath) {
+    const data = await bucket.file(gcsFilePath).download();
+    return data[0];
+  }
 
-      await writeFilePromise(diffPath, data.getBuffer());
+  /**
+   * Saves the golden hash
+   * @param {string} goldenHash The hash of the golden image
+   * @private
+   */
+  async saveGoldenHash_(goldenHash) {
+    const goldenFile = await readFilePromise(goldenFilePath);
+    const goldenJSON = JSON.parse(goldenFile);
+    goldenJSON[this.urlPath_] = goldenHash;
+    const goldenContent = JSON.stringify(goldenJSON, null, '  ');
+    await writeFilePromise(goldenFilePath, `${goldenContent}\r\n`);
+  }
 
-      assert.isBelow(Number(data.misMatchPercentage), 0.01);
+  /**
+   * Saves the given image to Google Cloud Storage with optional metadata
+   * @param {string} imagePath The path to the image
+   * @param {!Buffer} imageBuffer The image buffer
+   * @param {!Object=} customMetadata Optional custom metadata
+   * @private
+   */
+  async saveImage_(imagePath, imageBuffer, customMetadata={}) {
+    const metadata = Object.assign({}, defaultMetadata, customMetadata);
+    const file = bucket.file(imagePath);
+
+    // Check if file exists and exit if it does
+    const [exists] = await file.exists();
+    if (exists) {
+      console.log('✔︎ No changes to', imagePath);
       return;
+    }
+
+    // Create a new stream from the image buffer
+    let stream = new Readable();
+    stream.push(imageBuffer);
+    stream.push(null);
```

Comment:
Maybe add a comment on this line explaining that [`null` indicates EOF](https://nodejs.org/api/stream.html#stream_readable_push_chunk_encoding)